src/polar/kernels.jl

```diff
@@ -181,22 +192,74 @@ KA.@kernel function residual_adj_kernel!(F, adj_F, v_m, adj_v_m, v_a, adj_v_a,
     adj_pinj[fr] -= adj_F[i]
 end
 
-function residual_adj_polar!(F, adj_F, v_m, adj_v_m, v_a, adj_v_a,
+KA.@kernel function adj_residual_node_kernel!(F, adj_F, vm, adj_vm, va, adj_va,
+                                  colptr, rowval,
+                                  ybus_re_nzval, ybus_im_nzval,
+                                  edge_vm_from, edge_vm_to,
+                                  edge_va_from, edge_va_to,
+                                  pinj, adj_pinj, qinj, pv, pq)
+
+    npv = size(pv, 1)
+    npq = size(pq, 1)
+
+    i = @index(Global, Linear)
+    # REAL PV: 1:npv
+    # REAL PQ: (npv+1:npv+npq)
+    # IMAG PQ: (npv+npq+1:npv+2npq)
+    @inbounds for c in colptr[i]:colptr[i+1]-1
+        to = rowval[c]
+        adj_vm[i] += edge_vm_from[c]
+        adj_vm[i] += edge_vm_to[c]
+        adj_va[i] += edge_va_from[c]
+        adj_va[i] += edge_va_to[c]
+    end
+end
+
+function adj_residual_polar!(F, adj_F, vm, adj_vm, va, adj_va,
                          ybus_re, ybus_im,
                          pinj, adj_pinj, qinj, pv, pq, nbus)
     npv = length(pv)
     npq = length(pq)
+    nvbus = length(vm)
+    nnz = length(ybus_re.nzval)
+    T = eltype(vm)
+    edge_vm_from = zeros(T, nnz)
```

Comment:
Do you know if we could reuse these variables to compute the adjoint of the line flow constraints?

Reply:
Yes we could. I guess this belongs into the cache together with the adjoints.